src/completion.ml

```diff
@@ -0,0 +1,292 @@
+(** Completion procedure for closed equations *)
+
+open Terms
+open Basics
+open Extra
+open Timed
+
+type 'a cmp = 'a -> 'a -> int
+
+(** [ord_lex ord] computes the lexicographic order corresponding to the
+    alphabetical order [ord] *)
+let rec ord_lex : 'a cmp -> 'a list cmp = fun ord l1 l2 ->
+  match (l1, l2) with
+  | [], []             -> 0
+  | [], _              -> -1
+  | _, []              -> 1
+  | h1 :: t1, h2 :: t2 ->
+      match ord h1 h2 with
+      | 0 -> ord_lex ord t1 t2
+      | x -> x
+
+let get_sym : term -> sym = fun t -> match unfold t with
+  | Symb (s, _) -> s
+  | _           -> assert false
+
+(** [lpo ord] computes the lexicographic path ordering corresponding to
+    the strict total order [ord] on the set of all function symbols *)
+let rec lpo : sym cmp -> term cmp = fun ord t1 t2 ->
+  let f, args = get_args t1 in
+  let f = get_sym f in
+  if List.exists (fun x -> lpo ord x t2 >= 0) args then 1
+  else
+    let g, args' = get_args t2 in
+    let g = get_sym g in
+    match ord f g with
+    | 1 ->
+        if List.for_all (fun x -> lpo ord t1 x > 0) args' then 1
+        else -1
+    | 0 ->
+        if List.for_all (fun x -> lpo ord t1 x > 0) args' then
+          ord_lex (lpo ord) args args'
+        else -1
+    | _ -> -1
+
+(** [build_meta_type k] builds the type “∀(x₁:A₁) (x₂:A₂) ⋯ (xk:Ak), B” where
+    “x₁” through “xk” are fresh variables, “Ai = Mi[x₁,⋯,x(i-1)]”, “Mi” is a
+    new metavar of arity “i-1” and type “∀(x₁:A₂) ⋯ (x(i-1):A(i-1), TYPE”. *)
+let build_meta_type : int -> term = fun k ->
+  assert (k>=0);
+  let vs = Bindlib.new_mvar mkfree (Array.make k "x") in
+  let rec build_prod k p =
+    if k = 0 then p
+    else
+      let k = k-1 in
+      let mk_typ = Bindlib.unbox (build_prod k _Type) in
+      let mk = fresh_meta mk_typ k in
+      let tk = _Meta mk (Array.map _Vari (Array.sub vs 0 k)) in
+      let b = Bindlib.bind_var vs.(k) p in
+      let p = Bindlib.box_apply2 (fun a b -> Prod(a,b)) tk b in
+      build_prod k p
+  in
+  let mk_typ = Bindlib.unbox (build_prod k _Type) (*FIXME?*) in
+  let mk = fresh_meta mk_typ k in
+  let tk = _Meta mk (Array.map _Vari vs) in
+  Bindlib.unbox (build_prod k tk)
+
+type rule = sym * Terms.rule
+
+(** [to_m k metas t] computes a new (boxed) term by replacing every pattern
+    variable in [t] by a fresh meta-variable and store the latter in [metas],
+    where [k] indicates the order of the term obtained *)
+let rec to_m : int -> (meta option) array -> term -> tbox = fun k metas t ->
```

Comment:
Why having a meta option array and not a meta array?
Why the meta array is an input and not an output?

Reply:
When we consider a rule, the arity of rhs can be directly obtained by the function Bindlib.mbinder_arity. It is thus useful to take an empty array as an input when we apply to_m. Since we could have several occurrences of the same meta-variable in a term, we use a meta option array to prevent reinitialization.

Reply:
Ok, I see why the meta array must be given as input. But I think that you do not need the option, do you?

Reply:
We could even define the meta array before and not update it later as metavariable names are not very important.